debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/ParallelSnapshotReader.java

```diff
@@ -0,0 +1,191 @@
+     * @param config the current connector configuration.
+     * @param noSnapshotContext The context for those tables not undergoing a snapshot.
+     * @param snapshotFilters {@link Filters} matching the tables that should be snapshotted.
+     */
+    public ParallelSnapshotReader(Configuration config,
+                                  MySqlTaskContext noSnapshotContext,
+                                  Filters snapshotFilters) {
+        AtomicBoolean oldTablesReaderNearEnd = new AtomicBoolean(false);
+        AtomicBoolean newTablesReaderNearEnd = new AtomicBoolean(false);
+        ParallelHaltingPredicate oldTablesReaderHaltingPredicate =
+            new ParallelHaltingPredicate(oldTablesReaderNearEnd, newTablesReaderNearEnd);
+        ParallelHaltingPredicate newTablesReaderHaltingPredicate =
+            new ParallelHaltingPredicate(newTablesReaderNearEnd, oldTablesReaderNearEnd);
+
+        this.oldTablesReader = new BinlogReader("oldBinlog", noSnapshotContext, oldTablesReaderHaltingPredicate);
+
+        MySqlTaskContext newTablesContext = new MySqlTaskContext(config, snapshotFilters);
+        SnapshotReader newTablesSnapshotReader = new SnapshotReader("newSnapshot", newTablesContext);
+
+        this.newTablesBinlogReader = new BinlogReader("newBinlog", newTablesContext, newTablesReaderHaltingPredicate);
+        this.newTablesReader = new ChainedReader().add(newTablesSnapshotReader).add(newTablesBinlogReader);
+
+    }
+
+    // for testing purposes
+    /*package private*/ ParallelSnapshotReader(BinlogReader oldTablesBinlogReader,
+                                               SnapshotReader newTablesSnapshotReader,
+                                               BinlogReader newTablesBinlogReader) {
+        this.oldTablesReader = oldTablesBinlogReader;
+        this.newTablesBinlogReader = newTablesBinlogReader;
+        this.newTablesReader = new ChainedReader().add(newTablesSnapshotReader).add(newTablesBinlogReader);
+    }
+
+    /**
+     * Create and return a {@link ReconcilingBinlogReader} for the two binlog readers contained in this
+     * ParallelSnapshotReader.
+     * @return a {@link ReconcilingBinlogReader}
+     */
+    public ReconcilingBinlogReader createReconcillingBinlogReader() {
+        return new ReconcilingBinlogReader(oldTablesReader, newTablesBinlogReader);
+    }
+
+    @Override
+    public void uponCompletion(Runnable handler) {
+        uponCompletion.set(handler);
+    }
+
+    @Override
+    public void initialize() {
+        oldTablesReader.initialize();
+        newTablesReader.initialize();
+    }
+
+    @Override
+    public void start() {
+        if (running.compareAndSet(false, true)) {
+            oldTablesReader.start();
+            newTablesReader.start();
+        }
+    }
+
+    @Override
+    public void stop() {
+        try {
+            logger.info("Stopping the {} reader", oldTablesReader.name());
+            oldTablesReader.stop();
+        } catch (Throwable t) {
```

Comment:
I'm slightly confused by what you are asking here, can you elaborate?
I am logging the incoming exception.
If you are asking me to let the throwable percolate up, I can't, because the stop method I'm inheriting doesn't throw.
I probably shouldn't mark the reader as stopped if the sub-readers weren't stopped though.